### PR TITLE
Fix FileNotFoundExceptions when using symlinks

### LINKED ISSUE #6928: [Bug]Crash when detecting keys, with symlinks
### Description of the issue

The UI crashes upon open when attempting to detect the key files, when using symlinks instead of physical files.

### Reproduction steps

Create key files in system folder with symlinks vs physical files.
Close and reopen UI.

### Log file

[Ryujinx_1.1.1330_2024-06-10_10-04-06.log](https://github.com/user-attachments/files/15759696/Ryujinx_1.1.1330_2024-06-10_10-04-06.log)

### OS

Arch/Steam OS Holo

### Ryujinx version

1.1.1330

### Game version

_No response_

### CPU

_No response_

### GPU

_No response_

### RAM

_No response_

### List of applied mods

_No response_

### Additional context?

_No response_

### LINKED ISSUE #6983: [Bug] Symlinked folders are not correctly managed
### Description of the issue

Greetings,

Xargon from RetroDECK here.

We have an issue on our Ryujinx RetroDECK's "build".
For many compiling reasons we are using the approach of extracting the AppImage version of Ryujinx in our readonly filesystem partitions inside Flatpak (we were unable to build our own manifest so we fall back on this method for the moment).

```yml
  - name: Ryujinx
    buildsystem: simple
    build-commands:
      - |
        tar -zxvf *.tar.gz
        mv -v publish ${FLATPAK_DEST}/ryujinx
        ln -sv ${FLATPAK_DEST}/ryujinx/Ryujinx ${FLATPAK_DEST}/bin/Ryujinx.sh
    sources:
      - type: file
        url: https://github.com/Ryujinx/release-channel-master/releases/download/1.1.1330/ryujinx-1.1.1330-linux_x64.tar.gz
        sha256: 1e9158a3f55ff128e089ac3778dfb35127f2cab9d59bc52b5fe63c7b7ac2f6cb
```

This version is working in our `RetroDECK 0.8.1b`, Ryujinx version in this build is `1.1.1297`, but now we are releasing a newer version `0.8.2b` and we updated Ryujinx to `1.1.1330` (and probably we will update more before the actual release of this version as is taking time to finish up).

One of the main features of RetroDECK is that the `retrodeck` folder is portable: if the user backups or moves the `retrodeck` it folder should be seamless for all the emulators as they are finding all their needed data (games/saves/states/fir◯◯◯re) without any migration process: just place that folder on another computer and you're good to go.
To achieve this sometimes we can configure the emulators to point to our places, sometimes, if the emulators don't support this feature we just symlink the folder like in the case of Ryujinx:

- saves folder `bis` into `retrodeck/saves/switch/ryujinx/nand`
- `sdcard` into `retrodeck/saves/switch/ryujinx/sdcard`
- "fir◯◯◯re" folder `system` into `retrodeck/bios/sw◯◯ch/k◯◯s`
- actual fir◯◯◯re folder `bis/system/Contents/` into `retrodeck/bios/sw◯◯ch/fir◯◯◯re`

The symlinked folders are inside he Flatpak domain (we got permissions to write there), however this approach is now broken as Ryujinx cannot recognize the installed fir◯◯◯re or k◯◯s.
Some users even in `RetroDECK 0.8.[0,1]b` reported that this approach was failing to them, but resetting the emulator (with RetroDECK defaults) and sometimes reinstalling the fir◯◯◯re solved their issue, thing that is not happening to me in our `cooker` (test) build.

What I am asking here is if you can please check if there is a bug in this symlinking approach or, even better, if you could implement the feature to point to specific folders paths for fir◯◯◯re, k◯◯s, nand and saves in order to get rid of the symlinks completely in honor of a maybe cleaner approach.

Thanks a lot for your efforts and keep up with this amazing project, especially now.

---

**TL;DR:** symlinks are some times not working for the folder described above, can you please check for bugs or implement a way to edit the needed folders?

### Reproduction steps

- Download RetroDECK [cooker-0.8.2b-UrareUntax](https://github.com/XargonWan/RetroDECK-cooker/releases/tag/cooker-0.8.2b-UrareUntax) build, install it.
- Start RetroDECK once for the first setup
- place one game into the retrodeck/roms/s◯◯◯h folder
- place the k◯◯s in `retrodeck/bios/sw◯◯ch/k◯◯s`
- Open Ryujinx via RetroDECK Configurator and install the fir◯◯◯re

Then try this some times:
- Run a game
- Close everything (RetroDECK) and such

### Log file

(not a log but a terminal output sorry)

Xargon
xargonwan
Online
da: xargonwan 
mac gay (isaac) — 21/06/2024 05:05
other than that seems mostly ok
Immagine
rip in peri peri — 21/06/2024 05:10
there is a clear that is sporadically not working
it's a color clear on the main target
the sky is drawn with an addition, which is why it blows out completely (it repeatedly adds and adds across frames til it becomes white) 
mac gay (isaac) — 21/06/2024 05:10
Right
rip in peri peri — 21/06/2024 05:10
the cloud shadows should be limited to the land using stencil
mac gay (isaac) — 21/06/2024 05:11
So still stencil problems then...
rip in peri peri — 21/06/2024 05:11
i don't know why that isn't working - with stencil clears at least one component of that stencil tested draw is working, and it uses the same state as the later cloud shadow draw
mac gay (isaac) — 21/06/2024 05:11
weird
i would capture but :YEP:
yam — 21/06/2024 05:19
does the xcode crash trace have any clues
mac gay (isaac) — 21/06/2024 05:22
not rly
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Xcode [87529]
Path:                  /Applications/Xcode-16.0.0-Beta.app/Contents/MacOS/Xcode... (260 KB rimanenti)
Mostra
XcodeCrash.txt
310 KB
its some internal gputools seg fault 
rip in peri peri — 21/06/2024 05:24
GPUMTLDebuggerController setUpCaptureDataSources:stateMirrorDataSource:
somewhere in here
yam — 21/06/2024 05:25
yeah that's a clue of sorts
rip in peri peri — 21/06/2024 05:25
i guess it's some mechanism to copy gpu state for replay
mac gay (isaac) — 21/06/2024 05:26
nat already looked into that function i assume and didnt see anything outwardly sus
since it happens when its replying i assume it has a trace created at that point
trying to see if theres some way i can get that trace before everything explodes so i can actually send it to them
yam — 21/06/2024 11:49
interesting tidbit here @mac gay (isaac) - i just got the exact same capture crash, except mine was when i attempted to capture an openGL application (which has a very basic render path) 
Tipo di allegato: unknown
Xcode-2024-06-20-214706.ips
209.01 KB
maybe not directly helpful but also probably worth passing onto nat/whomever 
looks like the exact same function addresses too
Soup — 21/06/2024 12:50
i might be barking up the wrong tree but is this why i can't write on audio input buffers
Immagine
Xargon | RetroDECK — 21/06/2024 16:12
Sorry, me again, I think I found a weirdness/bug(?)

Basically the "needed files" are in place but Ryujinx cannot find them, I will post some more info.
Immagine
00:00:06.154 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.157 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.158 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
Mostra
message.txt
15 KB
Immagine
This is running on RetroDECK (so flatpak environment).
The folders are symlinked but I believe/hope it's not an issue, right? 
And same happens for the firmware, actually I was trying to reinstall the firmware for the same issue
Emma Alyx — 21/06/2024 16:17
I don't think symlinks are enough to get around flatpak's filesystem sandbox. that would be ridiculous
you'll have to give ryujinx the additional permissions necessary to access the files and directories the symlinks are pointing to
Xargon | RetroDECK — 21/06/2024 16:18
No that is "inside" the flatpak domain as that path is allowed 
RetroDECK can write on sdcard and home folder
that are on the SD Card, even the roms are there
mac gay (isaac) — 21/06/2024 18:06
Oh interesting
If you can get me the project I can probably submit that as a reproduction ig?
mac gay (isaac) — 21/06/2024 18:23
@Samuliak77 New nightmare compute function just dropped (NieR Automata)
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

Mostra
message.txt
42 KB
:c_stephen_uhhhhh: greaaatttt
Immagine
And then we crash i assume trying to do geo-on-compute
Immagine
Samuliak77 — 21/06/2024 18:40
Maybe use the threadgroup_barrier function?
mac gay (isaac) — 21/06/2024 18:46
Yeah that i added
trying to get shared memories working now
Okay that actually was easy
now the problem is just that function
Immagine
although idk why its says its not qualified
rip in peri peri — 21/06/2024 18:50
i mentioned this earlier
compute doesn't have in parameters
mac gay (isaac) — 21/06/2024 18:51
Yeah ik
rip in peri peri — 21/06/2024 18:51
so it shouldn't be included on function calls, and on the function definitions
mac gay (isaac) — 21/06/2024 18:52
functions can reference stuff in parameters though
so we cant just remove it everywhere
﻿
Xargon | RetroDECK — 22/06/2024 21:58
Flatpak aside: does Ryujinx supports the symlinking of those folders on linux? 
Xargon | RetroDECK — 22/06/2024 21:45
Sorry to bother you again, did someone knows if there is a bug regarding the symlinks of said folders?
Xargon | RetroDECK — 21/06/2024 16:19
that are on the SD Card, even the roms are there
Xargon | RetroDECK — 21/06/2024 16:18
RetroDECK can write on sdcard and home folder
Xargon | RetroDECK — 21/06/2024 16:18
No that is "inside" the flatpak domain as that path is allowed 
Xargon | RetroDECK — 21/06/2024 16:15
And same happens for the firmware, actually I was trying to reinstall the firmware for the same issue
Xargon | RetroDECK — 21/06/2024 16:14
This is running on RetroDECK (so flatpak environment).
The folders are symlinked but I believe/hope it's not an issue, right? 
Xargon | RetroDECK — 21/06/2024 16:14
Immagine
Xargon | RetroDECK — 21/06/2024 16:13
00:00:06.154 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.157 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.158 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
Mostra
message.txt
15 KB
Xargon | RetroDECK — 21/06/2024 16:12
Sorry, me again, I think I found a weirdness/bug(?)

Basically the "needed files" are in place but Ryujinx cannot find them, I will post some more info.
Immagine
Xargon | RetroDECK — 19/06/2024 22:52
Actually that's just a terminal output snippet
Xargon | RetroDECK — 19/06/2024 22:51
The firmware should be in order as I never touched it, if it broke it's weird. However I can try to reinstall it. Thanks
Xargon | RetroDECK — 19/06/2024 22:48
Greetings, 

Xargon from RetroDECK here.
As some of you already know RetroDECK is flatpacking Ryujinx as default switch emulator.

Recently we updated Ryujinx, llvm18 and the KDE Framework to 6.7, I wonder if the this error is related to one of these.

I tried only one game actually, but was a game that I was actually playing on Ryujinx.

Any help?
Thanks in advance
00:00:00.013 |N| Application ReloadConfig: Loading configuration from: /home/[redacted]/.var/app/net.retrodeck.retrodeck/config/Ryujinx/Config.json
00:00:00.428 |I| Configuration LogValueChange: EnableFileLog set to: True
00:00:00.502 |I| Configuration LogValueChange: ResScale set to: 1
00:00:00.513 |I| Configuration LogValueChange: ResScaleCustom set to: 1
00:00:00.513 |I| Configuration LogValueChange: MaxAnisotropy set to: -1
00:00:00.515 |I| Configuration LogValueChange: AspectRatio set to: Fixed16x9
Mostra
message1.txt
10 KB
Xargon | RetroDECK — 27/02/2024 18:30
Quick question: is it possible to install the firmware via cli?
Xargon | RetroDECK — 27/02/2024 18:30
Thanks to all your support now Ryujinx is correctly integrated in RetroDECK!  🎉
Xargon | RetroDECK — 16/02/2024 06:16
so I believe itś checkin other thigs as well
Xargon | RetroDECK — 16/02/2024 06:15
the second one is just /var/config inside flatpak
Xargon | RetroDECK — 16/02/2024 06:15
I want to set it in a state where it's already initialized
Xargon | RetroDECK — 16/02/2024 06:14
I tried to create that folder and put a Config.json but it told me that system folder was found and then not starting
Xargon | RetroDECK — 16/02/2024 06:14
/var/config?
Xargon | RetroDECK — 16/02/2024 02:58
How Ryujinx knows if it's a fresh install or not?
Xargon | RetroDECK — 15/02/2024 22:38
(I cannot write the "bottom" word)
Xargon | RetroDECK — 15/02/2024 22:38
Hi, I'm Xargon and this is Jackaxx 
Xargon | RetroDECK — 15/02/2024 22:37
I bet you never tested on Intel right? 🤣
Xargon | RetroDECK — 15/02/2024 22:35
Or maybe... can it be because my graphics card is just an intel chip? 
00:00:06.154 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.157 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.158 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #52483186)
00:00:06.246 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.MyValueTaskSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #52483186)
00:00:06.250 |E| UI : [Error] [IME] Error:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_78”
   at Tmds.DBus.Protocol.DBusConnection.MyValueTaskSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.IBus.IBusX11TextInputMethod.SetCapabilitiesCore(Boolean, Boolean)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.<>c__DisplayClass40_0.<<UpdateCapabilities>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Avalonia.FreeDesktop.DBusCallQueue.Process() (IBusX11TextInputMethod #52483186)
00:00:06.253 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_78”
   at Tmds.DBus.Protocol.DBusConnection.MyValueTaskSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.QueueOnErrorAsync(Exception e) (IBusX11TextInputMethod #52483186)
00:00:07.710 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #28630766)
00:00:07.710 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #28630766)
00:00:07.710 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #28630766)
00:00:07.712 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.MyValueTaskSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #28630766)
00:00:09.823 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #60677888)
00:00:09.823 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #60677888)
00:00:09.823 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #60677888)
00:00:09.825 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #60677888)
00:00:12.474 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #53754023)
00:00:12.474 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #53754023)
00:00:12.474 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #53754023)
00:00:12.488 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #62215569)
00:00:12.489 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #62215569)
00:00:12.489 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #62215569)
00:00:12.494 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #53754023)
00:00:12.499 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #62215569)
00:00:12.500 |E| UI : [Error] [IME] Error:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_82”
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.IBus.IBusX11TextInputMethod.SetCapabilitiesCore(Boolean, Boolean)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.<>c__DisplayClass40_0.<<UpdateCapabilities>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Avalonia.FreeDesktop.DBusCallQueue.Process() (IBusX11TextInputMethod #53754023)
00:00:12.502 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_82”
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.QueueOnErrorAsync(Exception e) (IBusX11TextInputMethod #53754023)
00:00:15.913 |E| Application HandleFirmwareInstallation: LibHac.Common.Keys.MissingKeyException: Unable to decrypt NCA section.
Key Type: Common
Key Name: key_area_key_application_11
   at LibHac.Tools.FsSystem.NcaUtils.Nca.GetDecryptedKey(Int32 index)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.GetContentKey(NcaKeyType type)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenAesCtrStorage(IStorage baseStorage, Int32 index, Int64 offset, UInt64 upperCounter)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenDecryptedStorage(IStorage baseStorage, Int32 index, Boolean decrypting)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenRawStorage(Int32 index, Boolean openEncrypted)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenRawStorage(Int32 index)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenStorage(Int32 index, IntegrityCheckLevel integrityCheckLevel, Boolean leaveCompressed)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenStorage(Int32 index, IntegrityCheckLevel integrityCheckLevel)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenFileSystem(Int32 index, IntegrityCheckLevel integrityCheckLevel)
   at LibHac.Tools.FsSystem.NcaUtils.Nca.OpenFileSystem(NcaSectionType type, IntegrityCheckLevel integrityCheckLevel)
   at Ryujinx.HLE.FileSystem.ContentManager.<VerifyFirmwarePackage>g__VerifyAndGetVersionZip|44_1(ZipArchive archive, <>c__DisplayClass44_0&) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.HLE/FileSystem/ContentManager.cs:line 746
   at Ryujinx.HLE.FileSystem.ContentManager.VerifyFirmwarePackage(String firmwarePackage) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.HLE/FileSystem/ContentManager.cs:line 654
   at Ryujinx.Ava.UI.ViewModels.MainWindowViewModel.HandleFirmwareInstallation(String filename) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx/UI/ViewModels/MainWindowViewModel.cs:line 1013
00:00:57.306 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #6421510)
00:00:57.306 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #6421510)
00:00:57.306 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #6421510)
00:00:57.308 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #6421510)
00:01:02.418 |E| UI : [Error] [FreeDesktopPlatform] OnNameChange failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #10378893)
00:01:02.418 |E| UI : [Error] [FreeDesktopPlatform] OnCommitText failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #10378893)
00:01:02.418 |E| UI : [Error] [FreeDesktopPlatform] OnForwardKey failed: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Tmds.DBus.Protocol.DBusConnection+Observer'. (IBusX11TextInputMethod #10378893)
00:01:02.421 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method Destroy is not implemented on interface org.freedesktop.IBus.Service
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.Dispose() (IBusX11TextInputMethod #10378893)
00:01:02.422 |E| UI : [Error] [IME] Error:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_84”
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.IBus.IBusX11TextInputMethod.SetCapabilitiesCore(Boolean, Boolean)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.<>c__DisplayClass40_0.<<UpdateCapabilities>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Avalonia.FreeDesktop.DBusCallQueue.Process() (IBusX11TextInputMethod #10378893)
00:01:02.424 |E| UI : [Error] [IME] Error while destroying the context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus/InputContext_84”
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync(MessageBuffer message)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync(MessageBuffer message)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.QueueOnErrorAsync(Exception e) (IBusX11TextInputMethod #10378893)

### OS

Linux

### Ryujinx version

1.1.1330

### Game version

_No response_

### CPU

_No response_

### GPU

_No response_

### RAM

_No response_

### List of applied mods

_No response_

### Additional context?

_No response_

### PR DESCRIPTION
**Pull request by TSRBerry**
This PR adds handling for symlinks when we are resolving paths. This should allow people to use symlinks within their Ryujinx directory without causing Ryujinx to crash because of that.

Since there are a lot of instances where files or directories are opened, I'm not 100% sure that I caught everything.
I couldn't find any issues during my limited testing, but I'd appreciate if a few other people could try this as well.
I know you can get pretty crazy with symlinks, but I only tested a few basic cases.

This PR depends on my multi-title PR, since I made a bunch of changes to the way files are loaded in there and I didn't want to redo this PR when that one gets merged.

---

Depends on #5638 
Fixes #6983
Fixes #6928

### CODE REVIEW

**Comment by XargonWan** [comment]
I tested it and it works great! Thanks a lot.
